### ISSUE TITLE
ROOT does not compile on musl libc

### ISSUE THREAD

**Issue by balejk** [root-project/root#9253]
I'm trying to compile ROOT v6.24/06 on Void Linux with musl libc, however the build fails with the following error message:
```cc
/builddir/root-6.24.06/core/clib/src/mmapsup.c:47:15: error: conflicting types for 'getpagesize'
   47 | extern size_t getpagesize PARAMS ((void));
      |               ^~~~~~~~~~~
In file included from /builddir/root-6.24.06/core/clib/res/mmprivate.h:56,
                 from /builddir/root-6.24.06/core/clib/src/mmapsup.c:26:
/usr/include/unistd.h:163:5: note: previous declaration of 'getpagesize' was here
  163 | int getpagesize(void);
      |     ^~~~~~~~~~~
```

The relevant part of `mmapsup.c` is:
```c
#if defined(R__LINUX) && !defined(R__GLIBC) && !defined(__CYGWIN__) \
   || (defined(__CYGWIN__) && (CYGWIN_VERSION_API_MAJOR > 0 || CYGWIN_VERSION_API_MINOR < 213))
extern size_t getpagesize PARAMS ((void));
#else
extern int getpagesize PARAMS ((void));
#endif
```

My understanding is that as musl doesn't define `R__GLIBC`, the condition resolves in favor of declaring `getpagesize` return value as `size_t`. However, this conflicts with the declaration present in `unistd.h` (which is included several lines above this), which declares it as `int` (in accordance with the POSIX standard).

The same code snippet also causes error during `core/clib/src/mvalloc.c` compilation.

It's not entirely clear to me why should `getpagesize` be declared in the first place, as it is declared in the included `unistd.h` file - my guess is for some kind of compatibility, although POSIX dropped it twenty years ago and Cygwin seems to have implemented it at least four years ago, likely more, I am not sure how this works on Windows - let alone as `size_t`. Thus my suggestion is to remove the above code block from both of the mentioned files entirely, which seems to fix the problem for me (at least until the compilation fails on another problem, which I haven't analyzed yet; and I'm still able to compile for glibc too) and my question is: would it break anything crucial (as I mentioned, `getpagesize` has been deprecated for more than twenty years now (at least on UNIX-like systems), so keeping support for some potentially obscure platforms which still expect it with `size_t` return value may not be of high importance (but again, I do not know what the situation is on Windows))? And if so, what is the right way to solve this?

Here's a [link](https://root-forum.cern.ch/t/compiling-error-conflicting-types-alpine-linux/28193) to a ROOT forum thread which describes the same problem (but does not provide any more information than is stated above).

**Comment by balejk** [comment]
Another compilation failure is caused by this error:
```gcc
[1/1065] Building CXX object core/base/CMakeFiles/Base.dir/src/TString.cxx.o                                                                                                                   
FAILED: core/base/CMakeFiles/Base.dir/src/TString.cxx.o                                                                                                                                        
/usr/bin/g++  -I/builddir/root-6.24.06/core/base/inc -I/builddir/root-6.24.06/build/builtins/pcre/PCRE-prefix/src/PCRE-build -I/builddir/root-6.24.06/core/base/res -I/builddir/root-6.24.06/co
re/foundation/res -I/builddir/root-6.24.06/core/clib/inc -I/builddir/root-6.24.06/core/meta/inc -I/builddir/root-6.24.06/core/cont/inc -I/builddir/root-6.24.06/core/thread/inc -I/builddir/roo
t-6.24.06/core/gui/inc -I/builddir/root-6.24.06/core/foundation/inc -I/builddir/root-6.24.06/build/ginclude -DNDEBUG -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -fdiagnost
ics-color=always -std=c++14 -Wno-implicit-fallthrough -Wno-noexcept-type -pipe  -Wshadow -Wall -W -Woverloaded-virtual -fsigned-char -pthread  -DNDEBUG -fPIC -std=c++14 -MD -MT core/base/CMak
eFiles/Base.dir/src/TString.cxx.o -MF core/base/CMakeFiles/Base.dir/src/TString.cxx.o.d -o core/base/CMakeFiles/Base.dir/src/TString.cxx.o -c /builddir/root-6.24.06/core/base/src/TString.cxx 
In file included from /builddir/root-6.24.06/core/base/src/TString.cxx:44:                                                                                                                     
/builddir/root-6.24.06/core/base/src/TString.cxx: In member function 'void TString::FormImp(const char*, __va_list_tag*)':                                                                     
/builddir/root-6.24.06/core/base/inc/Varargs.h:48:35: error: '__va_copy' was not declared in this scope; did you mean 'va_copy'?                                                                  48 | #     define R__VA_COPY(to, from) __va_copy((to), (from))                                                                                                                              
      |                                   ^~~~~~~~~                                                                                                                                            
/builddir/root-6.24.06/core/base/src/TString.cxx:2269:4: note: in expansion of macro 'R__VA_COPY'
 2269 |    R__VA_COPY(sap, ap);                                                                                                                                                                
      |    ^~~~~~~~~~                                                                                                                                                                          
/builddir/root-6.24.06/core/base/src/TString.cxx: In function 'char* SlowFormat(const char*, __va_list_tag*, int)':
/builddir/root-6.24.06/core/base/inc/Varargs.h:48:35: error: '__va_copy' was not declared in this scope; did you mean 'va_copy'?                                                               
   48 | #     define R__VA_COPY(to, from) __va_copy((to), (from))                         
      |                                   ^~~~~~~~~                                         
/builddir/root-6.24.06/core/base/src/TString.cxx:2366:4: note: in expansion of macro 'R__VA_COPY'
 2366 |    R__VA_COPY(sap, ap);
      |    ^~~~~~~~~~
/builddir/root-6.24.06/core/base/src/TString.cxx: In function 'char* Format(const char*, __va_list_tag*)':
/builddir/root-6.24.06/core/base/inc/Varargs.h:48:35: error: '__va_copy' was not declared in this scope; did you mean 'va_copy'?
   48 | #     define R__VA_COPY(to, from) __va_copy((to), (from))
      |                                   ^~~~~~~~~
/builddir/root-6.24.06/core/base/src/TString.cxx:2415:4: note: in expansion of macro 'R__VA_COPY'
 2415 |    R__VA_COPY(sap, ap);
      |    ^~~~~~~~~~
```
The relevant code is
```c
#if !defined(R__VA_COPY)
#  if defined(__GNUC__) && !defined(__FreeBSD__)
#     define R__VA_COPY(to, from) __va_copy((to), (from))
#  elif defined(__va_copy)
#     define R__VA_COPY(to, from) __va_copy((to), (from))
#  elif defined(va_copy)
#     define R__VA_COPY(to, from) va_copy((to), (from))
#  elif defined(_WIN32) && _MSC_VER < 1310
#     define R__VA_COPY(to, from) (*(to) = *(from))
#  else
#     define R__VA_COPY(to, from) ((to) = (from))
#  endif
#endif
```
I am actually not sure if this could be a bug in `gcc`, because [apparently](https://www.gnu.org/software/libc/manual/html_node/Argument-Macros.html) `__va_copy` should be defined by the compiler and thus should not(?) be affected by libc.
> These macros are no longer provided by the GNU C Library, but rather by the compiler. 

Anyway, is there any reason why the above code block could not be rewritten as
```c
#if !defined(R__VA_COPY)
#  if defined(va_copy) /* checking this first as va_copy is the one described by the C99 standard */
#     define R__VA_COPY(to, from) va_copy((to), (from))
#  elif defined(__va_copy)
#     define R__VA_COPY(to, from) __va_copy((to), (from))
#  elif defined(_WIN32) && _MSC_VER < 1310
#     define R__VA_COPY(to, from) (*(to) = *(from))
#  else
#     define R__VA_COPY(to, from) ((to) = (from))
#  endif
#endif
```
?

**Comment by balejk** [comment]
Next problem is that musl [apparently](https://www.gnu.org/software/gnulib/manual/html_node/fpu_005fcontrol_002eh.html) does not provide `fpu_control.h`. According to this [forum post](https://root-forum.cern.ch/t/compiling-error-conflicting-types-alpine-linux/28193/3), nothing from this file is in fact used and thus it should be possible to remove the include - based on the `fpu_control.h` [source code](https://code.woboq.org/userspace/glibc/sysdeps/x86/fpu_control.h.html) and a simple `grep`, this should hold for all ROOT components, except `math/mathcore/src/triangle.c`, which uses `_FPU_SETCW` at line 4888.

I am not familiar with what `fpu_control.h` does (apparently it has something to do with float precision), but according to [this](https://sourceforge.net/p/jamvm/patches/6/), it should be replaceable by a standard `fenv.h` header file, which I suggest to do (I have not tested it yet, but I will).

**Comment by Axel-Naumann** [comment]
Can you propose PRs for the items for which you identified solutions already? (And we don't support this libc and OS officially, but I bet your changes will actually do cleanup, not add complexity, so let's do it!)

**Comment by balejk** [comment]
Well, I wouldn't say I have identified solutions, I was hoping you would judge that, which is why I first opened an issue to discuss it.

Anyway, [here](https://github.com/root-project/root/pull/9260) is a pull request draft which comprises all the changes mentioned here. I was able to apply this patch to the current stable release and successfully build it against glibc. With musl there are still some other errors to be solved - I will try to address these later when I have the time.

**Comment by balejk** [comment]
```cc
FAILED: core/unix/CMakeFiles/Unix.dir/src/TUnixSystem.cxx.o                                                                                                                           [87/1897]
/usr/bin/g++  -I/builddir/root-6.24.06/core/unix/inc -I/builddir/root-6.24.06/core/clib/res -I/builddir/root-6.24.06/core/foundation/res -I/builddir/root-6.24.06/core/base/inc -I/builddir/roo
t-6.24.06/core/clib/inc -I/builddir/root-6.24.06/core/meta/inc -I/builddir/root-6.24.06/core/cont/inc -I/builddir/root-6.24.06/core/thread/inc -I/builddir/root-6.24.06/core/textinput/inc -I/b
uilddir/root-6.24.06/core/gui/inc -I/builddir/root-6.24.06/core/foundation/inc -I/builddir/root-6.24.06/build/ginclude -DNDEBUG -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2
 -fdiagnostics-color=always -std=c++14 -Wno-implicit-fallthrough -Wno-noexcept-type -pipe  -Wshadow -Wall -W -Woverloaded-virtual -fsigned-char -pthread  -DNDEBUG -fPIC -std=c++14 -MD -MT cor
e/unix/CMakeFiles/Unix.dir/src/TUnixSystem.cxx.o -MF core/unix/CMakeFiles/Unix.dir/src/TUnixSystem.cxx.o.d -o core/unix/CMakeFiles/Unix.dir/src/TUnixSystem.cxx.o -c /builddir/root-6.24.06/cor
e/unix/src/TUnixSystem.cxx                                                                     
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx: In member function 'virtual TInetAddress TUnixSystem::GetSockName(int)':
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3085:33: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3085 |    if (getsockname(sock, &addr, &len) == -1) {                         
      |                                 ^~~~                                                                                                                                                   
      |                                 |                                                      
      |                                 int*                                                   
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:                                                                                                                
/usr/include/sys/socket.h:338:52: note:   initializing argument 3 of 'int getsockname(int, sockaddr*, socklen_t*)'
  338 | int getsockname (int, struct sockaddr *__restrict, socklen_t *__restrict);
      |                                                    ^~~~~~~~~~~~~~~~~~~~~
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx: In member function 'virtual TInetAddress TUnixSystem::GetPeerName(int)':
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3121:33: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3121 |    if (getpeername(sock, &addr, &len) == -1) {                                                                                                                                         
      |                                 ^~~~                                                   
      |                                 |                                                      
      |                                 int*                                                                                                                                                   
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:           
/usr/include/sys/socket.h:339:52: note:   initializing argument 3 of 'int getpeername(int, sockaddr*, socklen_t*)'
  339 | int getpeername (int, struct sockaddr *__restrict, socklen_t *__restrict);
      |                                                    ^~~~~~~~~~~~~~~~~~~~~
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx: In member function 'virtual int TUnixSystem::GetSockOpt(int, int, int*)':
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3495:63: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3495 |       if (getsockopt(sock, SOL_SOCKET, SO_SNDBUF, (char*)val, &optlen) == -1) {
      |                                                               ^~~~~~~ 
      |                                                               |       
      |                                                               int*          
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:
/usr/include/sys/socket.h:348:50: note:   initializing argument 5 of 'int getsockopt(int, int, int, void*, socklen_t*)'
  348 | int getsockopt (int, int, int, void *__restrict, socklen_t *__restrict);                                                                                                               
      |                                                  ^~~~~~~~~~~~~~~~~~~~~
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3501:63: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3501 |       if (getsockopt(sock, SOL_SOCKET, SO_RCVBUF, (char*)val, &optlen) == -1) {
      |                                                               ^~~~~~~
      |                                                               |
      |                                                               int*
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:
/usr/include/sys/socket.h:348:50: note:   initializing argument 5 of 'int getsockopt(int, int, int, void*, socklen_t*)'
  348 | int getsockopt (int, int, int, void *__restrict, socklen_t *__restrict);
      |                                                  ^~~~~~~~~~~~~~~~~~~~~
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3507:66: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3507 |       if (getsockopt(sock, SOL_SOCKET, SO_OOBINLINE, (char*)val, &optlen) == -1) {
      |                                                                  ^~~~~~~
      |                                                                  |
      |                                                                  int*
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:
/usr/include/sys/socket.h:348:50: note:   initializing argument 5 of 'int getsockopt(int, int, int, void*, socklen_t*)'
  348 | int getsockopt (int, int, int, void *__restrict, socklen_t *__restrict);
      |                                                  ^~~~~~~~~~~~~~~~~~~~~
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3513:66: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3513 |       if (getsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, (char*)val, &optlen) == -1) {
      |                                                                  ^~~~~~~
      |                                                                  |
      |                                                                  int*
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:
/usr/include/sys/socket.h:348:50: note:   initializing argument 5 of 'int getsockopt(int, int, int, void*, socklen_t*)'
  348 | int getsockopt (int, int, int, void *__restrict, socklen_t *__restrict);
      |                                                  ^~~~~~~~~~~~~~~~~~~~~
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3519:66: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3519 |       if (getsockopt(sock, SOL_SOCKET, SO_REUSEADDR, (char*)val, &optlen) == -1) {
      |                                                                  ^~~~~~~
      |                                                                  |
      |                                                                  int*
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:
/usr/include/sys/socket.h:348:50: note:   initializing argument 5 of 'int getsockopt(int, int, int, void*, socklen_t*)'
  348 | int getsockopt (int, int, int, void *__restrict, socklen_t *__restrict);
      |                                                  ^~~~~~~~~~~~~~~~~~~~~
/builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:3525:66: error: invalid conversion from 'int*' to 'socklen_t*' {aka 'unsigned int*'} [-fpermissive]
 3525 |       if (getsockopt(sock, IPPROTO_TCP, TCP_NODELAY, (char*)val, &optlen) == -1) {
      |                                                                  ^~~~~~~
      |                                                                  |
      |                                                                  int*
In file included from /builddir/root-6.24.06/core/unix/src/TUnixSystem.cxx:101:
/usr/include/sys/socket.h:348:50: note:   initializing argument 5 of 'int getsockopt(int, int, int, void*, socklen_t*)'
  348 | int getsockopt (int, int, int, void *__restrict, socklen_t *__restrict);
      |                                                  ^~~~~~~~~~~~~~~~~~~~~
```
is the next error.

The problem stems from the fact, that the condition for using `socklen_t` is (among others) glibc:
```c
#if (defined(R__AIX) && !defined(_AIX43)) || \
    (defined(R__SUNGCC3) && !defined(__arch64__))
#   define USE_SIZE_T
#elif defined(R__GLIBC) || defined(R__FBSD) || \
      (defined(R__SUNGCC3) && defined(__arch64__)) || \
      defined(R__OBSD) || defined(MAC_OS_X_VERSION_10_4) || \
      (defined(R__AIX) && defined(_AIX43)) || \
      (defined(R__SOLARIS) && defined(_SOCKLEN_T))
#   define USE_SOCKLEN_T
#endif
```
[Here](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xsh_chap02.html#tag_22_02_10_06) is some context:
> All socklen_t types were originally (in BSD UNIX) of type int. During the development of POSIX.1-2017, it was decided to change all buffer lengths to size_t, which appears at face value to make sense. When dual mode 32/64-bit systems came along, this choice unnecessarily complicated system interfaces because size_t (with long) was a different size under ILP32 and LP64 models. Reverting to int would have happened except that some implementations had already shipped 64-bit-only interfaces. The compromise was a type which could be defined to be any size by the implementation: socklen_t.

I am not sure how to approach this, because the standardization of `socklen_t` seems fairly recent, but in the long run, avoiding this whole types dance altogether and simply using explicit `socklen_t` seems to make most sense and will make for a simpler code. I will appreciate any comments, especially regarding any compatibility issues this might cause. Else I will add this to the PR.
